Stream scope information only once

TimonPost opened the pull request:
This PR optimizes puffin bandwidth and scope times. This is achieved by only sending scope details once and via a separate stream. 

# Migration guide. 

Libraries using the profiling macros `profile_function` and `profile_scope!` don't have to change anything. This change is relevant and impactful for external libraries that register custom scopes manually. For example, timings gathered from GPU. 

A scope exists out of:
- metadata: function name, file path, line number, scope name
- Time information: start ms, end ms

Before this PR we streamed all information every time we recorded a scope to the puffin `GlobalProfiler`. After this PR, we only stream the metadata once the scope is first seen. Conceptually this makes a lot of sense but has some implications too users working with custom scopes not generated by the macros. Namely, a user library needs to register a scope once before reporting it to puffin. 

The main approach right now is the following:
- The sink is supposed to have a `FrameView`, a frame view stores a complete collection of all scope details that has been registered. 
- Scope meta-data can be requested via `ScopeId` (unique scope id for every scope). Scope meta-data is stored inside the `ScopeCollection` which every `FrameView` owns. 

```rust
fn print_scopes(frame_view: &FrameView) {
    let scope_collection = frame_view.scope_collection();

    for (name, scope_id) in scope_collection.scopes_by_name() {
        println!("{name:?} {scope_id:?}");
    }

    for (scope_id, scope_details) in scope_collection.scopes_by_id() {
        println!("{scope_id:?} {scope_details:?}");
    }
}

fn register_scopes_once(frame_view: &puffin::FrameView) {
    // This stores all scope details - can be used to find if scopes are already registered.
    let scope_collection = frame_view.scope_collection();

    // Those are the new custom scope details.
    let mut custom_scope_details = vec![
        puffin::ScopeDetails::from_scope_name("gpu")
                .with_file("gpu_file.rs")
                .with_line_nr(1),
        puffin::ScopeDetails::from_scope_name("gpu1")
                .with_file("gpu_file1.rs")
                .with_line_nr(1),
    ];

    // Only cal once for each new scope.
    // Use `scope_collection` to see if the scopes are new before registering.
    let new_scope_ids = puffin::GlobalProfiler::lock().register_user_scopes(&custom_scope_details);
}

fn report_custom_scopes() {
    // This is your frame view, can be the global frame view which will hook into the global profiler.
    // But can also be your own frame view with your own sink.
    let frame_view = puffin::GlobalFrameView::default();

    // Scopes should ONLY be registered once.
    // There are various ways to do this but is specific to your application.
    register_scopes_once(&frame_view.lock());

    // Custom assembled streams that capture time information for the registered scopes.
    let stream: puffin::Stream = _;

    // Report the custom user scopes.
    GlobalProfiler::lock().report_user_scopes(
        puffin::ThreadInfo {
            start_time_ns: None,
            name: "gpu".to_owned(),
        },
        &puffin::StreamInfo {
            num_scopes: 10,
            stream,
            depth: 5,
            range_ns: (0.0..10.0),
        }
        .as_stream_into_ref(),
    );

    // Iterate through all scopes, scopes are propagated from `GlobalProfiler` to the sink which writes them to the `FrameView`.
    print_scopes(&frame_view.lock());
}
```

# Benchmarks

In general a 8% speed up compared to the main on my mac with M1 max, about 10% on my computer AMD Ryzen 9 5900X. It uses roughtly about 70% less bandwidth.


```
profile_function        time:   [67.497 ns 67.811 ns 68.202 ns]
                        change: [-12.856% -10.578% -8.4281%] (p = 0.00 < 0.05)
profile_function_data   time:   [67.549 ns 68.003 ns 68.670 ns]
                        change: [-13.471% -11.663% -9.9536%] (p = 0.00 < 0.05)
profile_scope           time:   [67.110 ns 67.284 ns 67.473 ns]
                        change: [-7.5609% -6.5684% -5.4889%] (p = 0.00 < 0.05)
profile_scope_data      time:   [68.673 ns 69.036 ns 69.492 ns]
                        change: [-7.5932% -6.6513% -5.6345%] (p = 0.00 < 0.05)
profile_function_off    time:   [2.6392 ns 2.6413 ns 2.6438 ns]
                        change: [+11.028% +11.602% +12.444%] (p = 0.00 < 0.05)
profile_function_data_off
                        time:   [2.4242 ns 2.4272 ns 2.4318 ns]
                        change: [-7.5747% -7.0304% -6.4890%] (p = 0.00 < 0.05)
profile_scope_off       time:   [2.3711 ns 2.3782 ns 2.3856 ns]
                        change: [+0.0496% +0.5844% +1.1478%] (p = 0.04 < 0.05)
profile_scope_data_off  time:   [2.3753 ns 2.3835 ns 2.3920 ns]
                        change: [-9.3472% -8.5543% -7.9134%] (p = 0.00 < 0.05)
```

### Profile Locks

```
atomic set/get (criterion can not measure time)
thread_local_set        time:   [946.26 ps 950.81 ps 956.17 ps]
rw_lock_get             time:   [13.754 ns 13.788 ns 13.826 ns]
rw_lock_set             time:   [12.923 ns 12.962 ns 13.000 ns]
once_cell               time:   [315.48 ps 316.55 ps 317.65 ps]
once_lock               time:   [690.91 ps 694.27 ps 697.63 ps]
```

Ideally we want to use `once_cell` in the profiler macros however this seems to cause some deadlocks when running the profiler and I'm not sure why. It seems to run fine when just profiling things. We should be able to have a cell here as its thread local right? 

# Changes

- Remove `location` and `id` from the profile scope.
- Remove string parsing in macros, instead, this is now done when creating a new frame for the `GlobalProfiler`.
- Put `ScopeCollection` into `FrameView` such that each view contains its own collection of scopes. Its possible to get a full snapshot of all seen scope details to each sink.
- How to pass the scope details to the `GlobalProfiler`
I pass the profiler scope details via type array to the `GlobalProfiler`, I started off with encoding it into a separate stream, but this complicates parsing/reading logic and is not really needed. For the HTTP server we can just serialize the new scopes and send them to the user.
- Change scope id to be the incremental monolithic integer instead of a string of name.
- Add new HTTP server/client protocol header
- Upgrade protocol version

Fixes: #168

emilk commented on puffin/src/data.rs:
````diff
@@ -17,8 +15,8 @@
 //! ```ignore
-//!    ')'          byte       Sentinel
-//!    time_ns      i64        Time stamp of when scope finished
+//!    ')'          byte       Sentinel   
````
tip: tell your editor to trim trailing whitespace

emilk commented on puffin/src/lib.rs:
```diff
@@ -679,33 +962,44 @@ pub fn clean_function_name(name: &'static str) -> String {
 #[doc(hidden)]
 #[inline(never)]
-pub fn short_file_name(path: &'static str) -> String {
+pub fn short_file_name<'a>(path: &'a str) -> String {
```
I big reason we shortened the file name and function name is to reduce bandwidth. Now that we only log it once, we should consider sending the full file path and the full function name, leaving it to the UI to shorten it.

We should still strip the `::{{closure}}::{{closure}}::f` suffix upfront though,
and I suggest we also remove the path up to the crate name (the folder before the `src` folder), but don't do any further path ellision

TimonPost replied:
Yes this is now done by the UI, or i store it in `ScopeDetailsOwned` which is created once for each scope. It contains the parsed names and can be read by the UI. I thought we could keep this logic it in the `report` function but at the cost that maybe first scope would be a couple ns longer.

TimonPost replied:
I can probably return a subset of the string as a static ref string, I try to keep it fully static str so that allocations are prevented.

emilk reviewed (commented):
Very exciting!

The new integer id:s should also speed up the scope merging in the UI

TimonPost commented:
External libraries inserting custom scopes can do like this:

1. First insert the new scopes
2. Second report the scopes them selfes

```
 let mut custom_scope_details = vec![
    CustomScopeDetails::default().with_name("gpu").with_file("gpu_file.rs").with_line_nr(1),
    CustomScopeDetails::default().with_name("gpu_1").with_file("gpu_file_1.rs").with_line_nr(5),
];

let scope_details = GlobalProfiler::scope_details();
scope_details.insert_custom_scopes(&custom_scope_details);

GlobalProfiler::lock().report_custom_scopes(
    puffin::ThreadInfo {
        start_time_ns: None,
        name: "gpu".to_owned(),
    },
    &puffin::StreamInfo {
        num_scopes: 10,
        stream: scope_streams, // write scopes to stream
        depth: 5,
        range_ns: (gpu_frame_start_ns, gpu_frame_start_ns + gpu_time_accum),
    }
    .as_stream_into_ref(),
);
```

Libraries using macros don't have to change anything. This change for external libraries is introduced since now scope details should be registered once before reporting the scope.

TimonPost commented:
I opened for review, feel free to start commenting, I'm not completely done yet with streaming the scope details via HTTP server, but most logic is in place.

jnises commented on puffin/src/frame_data.rs:
```diff
@@ -570,11 +614,73 @@ impl FrameData {
+                read.read_exact(&mut serialized_scopes)
+                    .context("Can not deserialize scope details")?;
+
+                println!("{}", serialized_scope_len);
```
```suggestion
```

jnises commented on puffin/src/lib.rs:
```diff
@@ -809,31 +1200,26 @@ fn test_short_file_name() {
-                _INITITIALIZED.call_once(|| {
-                    _FUNCTION_NAME = $crate::current_function_name!().leak();
-                    _LOCATION = format!("{}:{}", $crate::current_file_name!(), line!()).leak();
+            static SCOPE_ID: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);
```
The `Once` should already be taking care of the synchronization right? So you could avoid the atomic here?

TimonPost replied:
The std::sync::Once itself doesn't handle synchronization of the data after the initialization. It ensures that the initialization code is executed only once right?

TimonPost replied:
I need to initialize this static but can not use unsafe to mutate it.

TimonPost replied:
oh maybe i should use OnceCell?

jnises replied:
`Once` docs say:
```
When this function returns, it is guaranteed that some initialization has run and completed (it might not be the closure specified). It is also guaranteed that any memory writes performed by the executed closure can be reliably observed by other threads at this point (there is a happens-before relation between the closure and code executing after the return).
```
So should be ok to do an unsafe write to a static?
As it was before.

Or, yes, `OnceCell`, or more likely `OnceLock` should work, and be cleaner.

TimonPost replied:
the problem is that unsafe is not allowed in macros as crates often disallow unsafe, and then every macro needs a comment.

repi replied:
yes needs to get rid of the `unsafe`, we can't currently use latest puffin library in our own project (!) because of this. which is obviously a requirement for us :smile: 
- #168

jnises replied:
Ok. I guess `OnceLock` would still be ok then. It's unsafe would be hidden in the standard library

TimonPost replied:
Using oncelock now, oncecell ended up in a deadlock or -stuck- for benchmarks

jnises commented on puffin/src/lib.rs:
```diff
@@ -852,29 +1238,26 @@ macro_rules! profile_function {
     ($id:expr) => {
         $crate::profile_scope!($id, "");
     };
     ($id:expr, $data:expr) => {
```
do we want to intern `data` as well?

TimonPost replied:
intern?

repi replied:
hmm interesting, logically it can be completely dynamic data, but typically it is semi static. so doing string interning on it also could be benefit for most cases.  though this is used a lot less so think skip it for this PR and change

jnises replied:
I thought `data` was the literal used for the scope. But I guess that is `id`. What happens to `id` now? do we ignore it?

TimonPost replied:
id is scope_name now and is only send once. I think we can follow up on the move to more static strings. This is something Niklas had a PR for before

jnises replied:
But what about the macro argument `id`? It looks to me like that is discarded? Isn't that supposed to be the name of the scope that is to be shown in the gui?

TimonPost replied:
Good point! I am not sending/sharing it atm, need to fix that.

jnises commented on puffin/src/lib.rs:
```diff
@@ -595,11 +715,9 @@ impl ProfilerScope {
     /// and this is a good way to enforce it.
```
update comment to match new behavior

jnises commented on puffin/src/lib.rs:
```diff
@@ -430,6 +455,14 @@ impl ThreadProfiler {
+static SCOPE_ID_TRACKER: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);
+
+fn fetch_add_scope_id() -> ScopeId {
+    let new_id = SCOPE_ID_TRACKER.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```
I guess this could be relaxed?
But it might not be hot enough to matter?

TimonPost replied:
i am not entirely sure, i can experience a bit but thought it would need to be quite restrict we don't want to have any duplicates.

emilk replied:
This can for sure be `Relaxed` - it has no dependency on other data. It can have a pretty big performance impact.

jnises commented on puffin/src/lib.rs:
```diff
@@ -430,6 +455,14 @@ impl ThreadProfiler {
+/// Incremental monolithic counter to identify scopes.
+static SCOPE_ID_TRACKER: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);
+
+fn fetch_add_scope_id() -> ScopeId {
```
```suggestion
fn make_next_scope_id() -> ScopeId {
```

jnises commented on puffin/src/frame_data.rs:
```diff
@@ -570,11 +614,73 @@ impl FrameData {
+                    u32::from_le_bytes(streams_compressed_length) as usize;
+                let compression_kind = read.read_u8()?;
+                let compression_kind = CompressionKind::from_u8(compression_kind)?;
+                let mut streams_compressed = vec![0_u8; streams_compressed_length];
```
would it be worth it to reuse the same vector for multiple reads?

jnises replied:
I guess this one cloudn't be reused. But perhaps the others?

jnises commented on puffin/src/lib.rs:
```diff
@@ -489,34 +533,81 @@ impl GlobalProfiler {
 
     /// Manually add frame data.
     pub fn add_frame(&mut self, new_frame: Arc<FrameData>) {
+        // for new_scope in new_frame.registered_scopes {
+        //     self.scope_details.read_by_id(&new_scope, |details| {
+
+        //     })
+        // }
+
```
```suggestion
```

emilk commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +447,32 @@ impl FrameData {
+            self.new_scopes.clone()
+        };
+
+        for new_scope_id in &scopes {
```
Does `scopes` and  `new_scopes` need to be a `HashSet`?

I don't see any queries being done of it. Using just a `Vec<ScopeId>` would be a lot faster, and give consistent order during the loop here

emilk commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +447,32 @@ impl FrameData {
+                        scope_name: details.scope_name.to_string(),
+                        function_name: details.dynamic_function_name.to_string(),
+                        file_path: details.dynamic_file_path.to_string(),
+                        line_nmr: details.line_nr,
```
inconsistent naming

```suggestion
                        line_nr: details.line_nr,
```

emilk commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +447,32 @@ impl FrameData {
 
+        let mut to_serialize_scopes = Vec::new();
+
+        let scopes = if write_all_scopes {
```
Nit: this `bool` is called `send_all_scopes` in other places

emilk commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +447,32 @@ impl FrameData {
+
+        let scopes = if write_all_scopes {
+            scope_details
+                .scopes_by_id(|all_scopes| all_scopes.keys().cloned().collect::<HashSet<ScopeId>>())
```
The keys should be `Copy`

```suggestion
                .scopes_by_id(|all_scopes| all_scopes.keys().copied().collect::<HashSet<ScopeId>>())
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+    /// Shorter variant of the raw function name.
+    /// This is more descriptive and shorter.
+    /// In the case of custom scopes provided scopes this contains the `scope name`.
+    pub dynamic_function_name: Cow<'static, str>,
```
`dynamic_` is weird choice of prefix given the definition. Maybe `readable_` or `shortened_` ?

emilk commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +447,32 @@ impl FrameData {
+                    to_serialize_scopes.push(crate::scope_details::SerdeScopeDetails {
+                        scope_id: *new_scope_id,
+                        scope_name: details.scope_name.to_string(),
+                        function_name: details.dynamic_function_name.to_string(),
+                        file_path: details.dynamic_file_path.to_string(),
```
We're serializing the shortened function names and file path here - is this intentional?

TimonPost replied:
Yes think so? Seems like the client / ui is most interested in the cleaned version of those strings rather then the raw unedited version.

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+use crate::{clean_function_name, fetch_add_scope_id, short_file_name, ScopeId};
+
+#[derive(Default, Clone)]
+struct InnerDetails {
+    pub(crate) scope_id_to_details: std::collections::HashMap<ScopeId, ScopeDetailsOwned>,
+    pub(crate) string_to_scope_id: std::collections::HashMap<String, ScopeId>,
+}
+
+/// Provides read access to scope details.
+#[derive(Default, Clone)]
+pub struct ScopeDetails(
```
This is confusing naming.

`ScopeDetailsOwned`, `ScopeDetailsRef`, `CustomScopeDetails` is all about the details of a _single_ profile scope.

…but `ScopeDetails` and `InnerDetals` is the collection of all scopes.

emilk replied:
This is also an argument against `ScopeDetails` as a name: it can't be pluralized. "The details for one scope vs the details of many scopes". Better then with `ScopeInfo`: "The info of one scope vs the info**s** for many scopes.

So I would suggest`ScopeInfo` for info about a single scope, and then `AllScopeInfos` (to be super explicit) for this collection of all infos we have here.

emilk commented on puffin/src/lib.rs:
```diff
@@ -612,6 +715,14 @@ impl Drop for ProfilerScope {
     }
 }
 
+/// A unique id for each scope and `ScopeInfo`.
```
You should use proper ref-links. Then the compiler could have told you of your mistake here. It also leads to better docs.

```suggestion
/// A unique id for each scope and [`ScopeDetails`].
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+    pub scope_id: ScopeId,
+    // Custom provided static id that identifiers a scope in a function.
+    pub scope_name: &'static str,
+    /// Scope name, or function name (previously called "id")
+    pub raw_function_name: &'static str,
```
Wait, is this sometime the scope name given to `profile_scope!` ? Then `raw_function_name` is quite confusing

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+/// Scope details that are registered by the macros.
+/// This is only used internally and should not be exposed.
+#[derive(Debug, Clone, Copy, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct ScopeDetailsRef {
```
We should make it clear (by name and docstring) that this is the static information that doesn't change with each invocation.

Perhaps `StaticScopeInfo` would be a good name?

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+        }
+    }
+
+    pub fn with_function_name(mut self, name: impl Into<Cow<'static, str>>) -> Self {
```
It is very important that all public small functions like these are marked `#[inline]`

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+/// Custom scope details that can be registered by external users.
+/// Instantiate this type once for each custom scope that you record manually.
+/// Custom scopes can be registered via `GlobalProfiler::scope_details().insert_scopes()`.
+// This type provides slightly more convenient api for external users.
+#[derive(Debug, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+pub struct CustomScopeDetails {
```
Why not reuse `ScopeDetailsRef`? It can be made as ergonomic with the same builder functions, no? I guess there will be a slight overhead with the `Cow` ?

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+    /// In the case of custom scopes provided scopes this contains the `file name`.
+    pub dynamic_file_path: Cow<'static, str>,
+    /// The full raw file path to the file in which the scope was located.
+    /// This is empty for custom scopes.
+    pub raw_file_path: &'static str,
```
How can we use `&'static` in a struct `ScopeDetailsOwned`? Aren't the raw file paths logged?

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+
+#[derive(Debug, Default, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+/// This struct contains scope details and can be read by external libraries.
+pub struct ScopeDetailsOwned {
```
So this is what we deserialize into?

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+        ScopeDetailsOwned {
+            scope_name: value.scope_name.clone(),
+            dynamic_function_name: value.function_name.clone(),
+            raw_function_name: "-", // user provided scopes are non-static
```
Again, let's use the Rust type system and use `None`

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,212 @@
+    feature = "serialization",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub(crate) struct SerdeScopeDetails {
```
This is the fourth copy now - this seems silly. Surely we can unify a bit!

For instance, we could design this so that `ScopeDetailsRef` can be converted to `CustomScopeDetails`, and then implement serde on `CustomScopeDetails`

emilk reviewed (changes requested):
This is a good start, but I find the five different kinds of `ScopeDetails` quite confusing, including their fields.

TimonPost commented:
Yes the different scope details are a bit of a pain I'm not sure how to resolve. 

- `CustomScopeDetails`: Is user-facing, with both options for dynamic and static strings. 
- `ScopeDetailsOwned`: Is internal-facing, it has hard-clones and heap-allocated strings and is meant for external apps to read. It contains some of the cleanups of function names and file paths. 
- `ScopeDetailsSerde`: Is serde focused, 'static str made some borrow issues regarding deserialization, cow can not be serialized. And we might not want to send all cleaned-up strings to reduce bandwidth. S
- `ScopeDetailsRef`: This one is internally focused and optimized for speed as it only contains static refs and copy data. It should be a cheap data structure. 

I will have a look and pay some extra time unifying those usecases while keeping in mind the usecase for static strings vs heap strings.

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,261 @@
+    /// Unique scope identifier.
+    // Always initialized once registered.
+    // It is `None` when an external library has yet to register this scope.
+    pub(crate) scope_id: Option<ScopeId>,
```
If we use `ScopeId(pub NonZeroU32)` then this field will shrink in size, thanks to `enum niche optimization`

TimonPost commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,261 @@
+struct Inner {
+    // Store a both-way map, memory wise this can be a bit redundant but allows for faster access of information by external libs.
+    pub(crate) scope_id_to_details: std::collections::HashMap<ScopeId, ScopeDetails>,
+    pub(crate) string_to_scope_id: std::collections::HashMap<String, ScopeId>,
```
This is a feature I added to support external creates. External crates might register scopes manually and when they want to lookup scope details might want to do this by name rather then scope id. We can return new scope id's for when a library registers scopes with `register_custom_scope` and then the external library should map those to the custom scope names. I thought perhaps we could do this internally and make this flow slightly easier for the users. 

What you think?

emilk replied:
Thanks for the added context!

TimonPost commented on puffin/src/lib.rs:
```diff
@@ -525,13 +630,34 @@ impl GlobalProfiler {
+        std::mem::take(&mut self.scope_delta)
+    }
+
+    pub fn scope_collection() -> ScopeCollection {
+        let lock = Self::lock();
+        lock.scope_collection.clone()
+    }
```
There are some issues with a sink-only approach, which I would prefer too. For example, in case of web-server. Webserver profiles, I connect with client, I'm adding a sync for this client on the server, I wont get any old deltas. The global profiler is the only place it seems I can objectively store what scope details we have and can provide them to any new sinks registered after the first scopes have been registered.

TimonPost commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,261 @@
+/// Provides fast read access to scope details.
+/// This collection can be cloned safely.
+#[derive(Default, Clone)]
+pub struct ScopeCollection(Arc<RwLock<Inner>>);
```
Removing this will result in having to always have to lock the `GlobalProfiler` code even if the UI/userspace just wants to do a quick lookup of scope details. I thought it could be nice to have a read lock here as 99% of the usecases will involve read-only access.

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,285 @@
+//
+// 1. Almost every access only requires read access so there is no need for mutex lock the global profiler.
+// 2. Its important to guarantee that the profiler path is lock-free.
+// 2. External libraries like the http server or ui require read/write access to scopes.
+// But that can easily end up in deadlocks if a profile scope is executed while scope details are being read.
```
I think the second second point here indicates the problem I've been trying to point out:

The scope _producer_ (processes using `profile_scope!` etc) need a global place to add scopes too.
But reusing this same global collection on the _consumer_ side (`puffin_egui`, `puffin_viewer`) seems like a mistake to me. It may very well be convenient (globals often are), but leads to things like the above mentioned deadlock.

Furthermore, any`profile_scope!` in `puffin_egui` or `puffin_viewer` will be writing over the same scopes.
`puffin_viewer` has a introspection mode where it profiles itself. This PR breaks it in two ways:
A) the scope info is wrong
B) it deadlocks after a while

![image](https://github.com/EmbarkStudios/puffin/assets/1148717/09e96ab4-9b80-4bda-a486-172fec8cfbe4)

On `main` all the data that `puffin_viewer` views is contained in a `FrameView`. It seems to me `FrameView` should contain a `ScopeCollection`.

TimonPost replied:
I managed to move it to `FrameView`, it comes with a little complexity for our users who insert custom scopes. They will have to use a `FrameView`, and make sure to get a full snapshot of scope details if the frame view was created after the scopes are registered.

TimonPost replied:
For convenience I can expose read access to the global scope collection from the `GlobalProfiler`, or would you like to keep that also restricted to `FrameView`. It is possible to do everything via a frame view

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,261 @@
+        ];
+        deserializer.deserialize_struct(stringify!(ScopeDetails), FIELDS, ScopeDetailsOwnedVisitor)
+    }
+}
```
Again, how is this different from `#[derive(serde::Serialize)]` ?

TimonPost replied:
Cow doesnt imple serde so need to do a manual implementation?

TimonPost replied:
Ah before I had statics here and it was not working well with deserialize trait bu now I can just use the default impl

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -711,30 +735,53 @@ fn to_ms(ns: NanoSecond) -> f64 {
     info: &Info,
     options: &mut Options,
     stream: &Stream,
+    scope_collection: &ScopeCollection,
```
you could also put `scope_collection` in `Info` to reduce the number of arguments being passed around

emilk commented on puffin_egui/src/lib.rs:
```diff
@@ -522,13 +534,12 @@ impl ProfilerUi {
+                .map(|frame| frame.unpacked())
+                .filter_map(|unpacked| unpacked.ok())
+                .collect();
+            SelectedFrames::try_from_vec(&frame_view.scope_collection, unpacked.clone())
```
this close seems unnecessary?

emilk commented on puffin_egui/src/lib.rs:
```diff
@@ -620,7 +645,7 @@ impl ProfilerUi {
                         let slowest_visible = self.show_frame_list(
                             ui,
-                            frame_view,
+                            frame_view.as_mut().unwrap(),
```
Don't `unwrap` - unwrap panics on failure

TimonPost replied:
some left overs from initial prototype can be made non mut

emilk commented on puffin_egui/src/lib.rs:
```diff
@@ -698,7 +723,7 @@ impl ProfilerUi {
         &mut self,
         ui: &mut egui::Ui,
-        frame_view: &FrameView,
+        frame_view: &mut FrameView,
```
does this really need to be `mut`?

emilk commented on puffin/src/lib.rs:
```diff
@@ -430,6 +476,17 @@ impl ThreadProfiler {
+static SCOPE_ID_TRACKER: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(1);
+
+fn fetch_add_scope_id() -> ScopeId {
+    let new_id = SCOPE_ID_TRACKER.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```
```suggestion
    let new_id = SCOPE_ID_TRACKER.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
```

TimonPost replied:
Its only called once I suppose so not that biggy, but yes can be relaxed

emilk commented on puffin/src/lib.rs:
```diff
@@ -447,15 +504,27 @@ pub struct GlobalProfiler {
 
     next_sink_id: FrameSinkId,
     sinks: std::collections::HashMap<FrameSinkId, FrameSink>,
+    propagate_all_scope_details: bool,
```
what is this? docstring please

Is this the same thing that is called `send_all_scopes` and `full_delta` in other places of the codebase?

emilk commented on puffin/src/lib.rs:
```diff
@@ -447,15 +504,27 @@ pub struct GlobalProfiler {
     sinks: std::collections::HashMap<FrameSinkId, FrameSink>,
+    propagate_all_scope_details: bool,
+
+    // When scope details are reported by the macros.
+    new_scopes: Vec<ScopeDetails>,
+    // When scope details are reported by external libraries.
+    new_custom_scopes: Vec<Arc<ScopeDetails>>,
```
Can these be combined into one? Using `Arc` for both seems like a pretty small price to pay for simpler code (one extra allocation for each new scope)

EDIT: we do that `Arc::new` in new_frame anyways

TimonPost replied:
Yes, think was that one has the scope id set and the other not. but i can check if scope id is none instead.

emilk commented on puffin/src/lib.rs:
```diff
@@ -489,34 +557,101 @@ impl GlobalProfiler {
+    /// Insert custom scopes into puffin.
+    /// Scopes details should only be registered once for each scope and need be inserted before being reported to puffin.
+    /// This function should only be relevant when your not using puffin through the profiler macros.
+    pub fn register_custom_scopes(&mut self, scopes: &[ScopeDetails]) {
```
nit: maybe "Use scope" is better nomenclature?

TimonPost replied:
not sure what you mean?

emilk replied:
oops, typo - I meant to write "User scope", i.e. replace "custom" with "user", as in the user generated them 🤷 

Just a suggestion, I have no strong preference

emilk commented on puffin/src/lib.rs:
```diff
@@ -489,34 +557,101 @@ impl GlobalProfiler {
 
+    /// Insert custom scopes into puffin.
+    /// Scopes details should only be registered once for each scope and need be inserted before being reported to puffin.
+    /// This function should only be relevant when your not using puffin through the profiler macros.
```
```suggestion
    /// This function is relevant when you're registering measurement not performed using the puffin profiler macros.
```

emilk commented on puffin/src/lib.rs:
```diff
@@ -489,34 +557,101 @@ impl GlobalProfiler {
     }
 
+    /// Insert custom scopes into puffin.
+    /// Scopes details should only be registered once for each scope and need be inserted before being reported to puffin.
```
This docstring should also document that the `scope_id`: s for the scopes are expected to be `None`

emilk commented on puffin/src/lib.rs:
```diff
@@ -623,16 +775,18 @@ pub fn type_name_of<T>(_: T) -> &'static str {
 #[inline(never)]
-pub fn clean_function_name(name: &'static str) -> String {
+pub fn clean_function_name(name: &str) -> String {
+    let Some(name) = name.strip_suffix(USELESS_SCOPE_NAME_POSTFIX) else {
```
nit: we should use the standard name (which the stdlib uses)

```suggestion
    let Some(name) = name.strip_suffix(USELESS_SCOPE_NAME_SUFFIX) else {
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,261 @@
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<ScopeId, Arc<ScopeDetails>>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.scope_id_to_details)
+    }
```
No need for the convoluted `impl FnMut` callbacks here. Just return a reference

emilk reviewed (commented):
getting closer!

TimonPost commented:
Some changes based on reviews:
- Removed the manual serde impl for the strucs as it wasn't needed anymore
- Some more docs and improved naming.
- Introduced `ScopeType` with `Function` and `Scope` variants to make a clearer distinction between what the name of a scope is. So we don't mix scope name / function name but make it more explicit to the caller what the scope name is about.

emilk commented on puffin_egui/src/stats.rs:
```diff
@@ -65,13 +71,23 @@ pub fn ui(ui: &mut egui::Ui, options: &mut Options, frames: &[std::sync::Arc<Unp
+                    ui.label(format!("{:?}", scope_details.function_name));
+
+                    if let Some(name) = &scope_details.scope_name {
+                        ui.label(format!("{}", name));
```
```suggestion
                        ui.label(name);
```

emilk commented on puffin_egui/src/stats.rs:
```diff
@@ -65,13 +71,23 @@ pub fn ui(ui: &mut egui::Ui, options: &mut Options, frames: &[std::sync::Arc<Unp
+                    if let Some(name) = &scope_details.scope_name {
+                        ui.label(format!("{}", name));
+                    } else {
+                        ui.label(format!("-"));
```
clippy should catch this…

```suggestion
                        ui.label("-");
```

emilk commented on puffin_egui/src/lib.rs:
```diff
@@ -659,7 +684,7 @@ impl ProfilerUi {
                 self.show_frame_list(
                     ui,
-                    frame_view,
+                    frame_view.as_mut().unwrap(),
```
unwrap=crash=bad

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -731,10 +753,14 @@ fn paint_scope(
-                ui.monospace(format!("id:       {}", scope.record.id));
-                if !scope.record.location.is_empty() {
-                    ui.monospace(format!("location: {}", scope.record.location));
+                ui.monospace(format!("id:       {}", scope_details.function_name));
```
mismatch between key (`id`) and value (`function_name`). This should probably be something like

```suggestion
                ui.monospace(format!("name:    {}", scope_details.name()));
```

where `ScopeDetails::name` either returns the named passed to `profile_scope!`, or the function name, or whatever is non-empty basically (e.g. for custom/user scopes)

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -731,10 +753,14 @@ fn paint_scope(
+                ui.monospace(format!("id:       {}", scope_details.function_name));
+                if !scope_details.file_path.is_empty() {
+                    ui.monospace(format!(
+                        "location: {}:{}",
+                        scope_details.file_path, scope_details.line_nr
```
```suggestion
                        "location: {}", scope_details.location()
```

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -731,10 +753,14 @@ fn paint_scope(
         }
 
         if result == PaintResult::Hovered {
+            let scope_details = info.scope_collection.fetch_by_id(&scope.id).unwrap();
```
unwrap = crash = bad

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -665,13 +679,21 @@ fn paint_record(
-        let duration_ms = to_ms(record.duration_ns);
-        let text = if record.data.is_empty() {
-            format!("{}{} {:6.3} ms {}", prefix, record.id, duration_ms, suffix)
+        let scope_details = info.scope_collection.fetch_by_id(&scope_id).unwrap();
```
unwrap = crash = … you know :)

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -665,13 +679,21 @@ fn paint_record(
-            format!("{}{} {:6.3} ms {}", prefix, record.id, duration_ms, suffix)
+        let scope_details = info.scope_collection.fetch_by_id(&scope_id).unwrap();
+
+        let scope_type = match scope_details.scope_type() {
+            ScopeType::Function(name) | ScopeType::Scope(name) => name,
+        };
```
build this in to the `ScopeType` instead? 🤷 

```suggestion
        let scope_type = match scope_details.name();
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,223 @@
+    /// Useful for fetching scope id by a static scope name.
+    /// For profiler scopes and custom scopes this is the manual provided name.
+    /// For function profiler scopes this is the function name.
+    pub fn scopes_by_name(&self) -> &HashMap<ScopeType, ScopeId> {
+        &self.0.type_to_scope_id
+    }
+
+    /// Fetches all registered scopes.
+    /// Useful for fetching scope details by a scope id.
+    pub fn scopes_by_id(&self) -> &HashMap<ScopeId, Arc<ScopeDetails>> {
```
`#[inline]`

emilk commented on puffin/src/lib.rs:
```diff
@@ -782,6 +938,78 @@ fn test_short_file_name() {
+    a();
+
+    GlobalProfiler::lock().new_frame();
+}
```
Nice test!

emilk commented:
Remember to make the UI nice too:

![image](https://github.com/EmbarkStudios/puffin/assets/1148717/c78c1a30-6840-48d5-8095-979a92500011)

jnises commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +466,21 @@ impl FrameData {
         write.write_u8(packed_streams.compression_kind as u8)?;
         write.write_all(&packed_streams.bytes)?;
 
+        let mut to_serialize_scopes = Vec::new();
```
you could reserve this to avoid some allocations.

jnises commented on puffin/src/frame_data.rs:
```diff
@@ -428,6 +466,21 @@ impl FrameData {
+        let mut to_serialize_scopes = Vec::new();
+
+        if send_all_scopes {
+            for scope in scope_collection.scopes_by_id().iter() {
```
perhaps this would work?
```suggestion
to_serialize_scopes.extend(scope_collection.scopes_by_id().values().copied());
```

jnises commented on puffin/src/lib.rs:
```diff
@@ -612,29 +737,45 @@ impl Drop for ProfilerScope {
+
+impl ScopeId {
+    #[cfg(test)]
+    pub(crate) fn new_unchecked(id: u32) -> Self {
```
doesn't seem so _un_-checked

jnises commented on puffin/src/lib.rs:
```diff
@@ -782,6 +934,78 @@ fn test_short_file_name() {
+// The macro defines 'f()' at the place where macro is called.
+// This code is located at the place of call and two closures deep.
+// Strip away this useless postfix.
+static USELESS_SCOPE_NAME_SUFFIX: &str = "::{{closure}}::{{closure}}::f";
```
any reason why this isn't `const`?

jnises commented on puffin/src/lib.rs:
```diff
@@ -859,31 +1071,27 @@ macro_rules! profile_function {
+        $crate::profile_scope!($name, "");
     };
-    ($id:expr, $data:expr) => {
+    ($name:expr, $data:expr) => {
```
should `name` be `literal` rather than `expr` here?
if `expr` allows variable expressions here things might get confusing?

jnises replied:
Ok. I can see how that would be useful.
Perhaps we should add to the doccomment that we only use the first invocation of the expression?

TimonPost replied:
I added it because in our large code base it seems someplace require it to be an expr and it relaxes the constraints a bit on how names are generated and were they come from.

TimonPost commented:
I will follow up sometime after this PR to update and add lots of clippy checks so we can automize that. I processed your commend and improved the popup windows a bit. I'll try make the width smaller.

<img width="922" alt="image" src="https://github.com/EmbarkStudios/puffin/assets/19969910/082fc1b5-05de-462d-adfa-601c635eeec4">

<img width="790" alt="image" src="https://github.com/EmbarkStudios/puffin/assets/19969910/2417def4-4278-4c2b-80d2-8c44b28aee97">

<img width="719" alt="image" src="https://github.com/EmbarkStudios/puffin/assets/19969910/3defc0b2-8233-4fb2-b41b-41a2b43ff034">

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -782,60 +819,94 @@ fn paint_merge_scope(
+                "duration:\t{:.3} ms",
                 to_ms(merge.duration_per_frame_ns)
             ));
         } else {
             ui.monospace(format!("sum of {} scopes", merge.num_pieces));
             ui.monospace(format!(
-                "total: {:7.3} ms",
+                "total:\t{:.3} ms",
                 to_ms(merge.duration_per_frame_ns)
             ));
             ui.monospace(format!(
-                "mean:  {:7.3} ms",
+                "mean:\t{:.3} ms",
                 to_ms(merge.duration_per_frame_ns) / (merge.num_pieces as f64),
             ));
-            ui.monospace(format!("max:   {:7.3} ms", to_ms(merge.max_duration_ns)));
+            ui.monospace(format!("max:\t{:.3} ms", to_ms(merge.max_duration_ns)));
```
Are you sure `\t` aligns subsequent labels? Have you tested the UI?

TimonPost replied:
see printscreens above, does that look correct to you?

emilk replied:
I don't see any screenshots with total/mean/max, so I made one:

![image](https://github.com/EmbarkStudios/puffin/assets/1148717/39072741-82bb-4f8c-9744-01659243f647)

and so no, it doesn't

TimonPost replied:
the current format is a bit confusing to me - manual spaces with spaces specified by {}. Had a couple of issues with it but I'll try revert back to that

emilk reviewed (changes requested):
Please take care to test the ui thoroughly

TimonPost commented on puffin/src/frame_data.rs:
```diff
@@ -570,6 +629,48 @@ impl FrameData {
+                    full_delta: false,
+                }))
+            } else if &header == b"PFD4" {
+                // Added 2023-12-01: Split up stream scope details from the record stream.
```
i'll update once ou approve

TimonPost commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -712,38 +748,39 @@ fn to_ms(ns: NanoSecond) -> f64 {
         let mut num_children = 0;
         for child_scope in Reader::with_offset(stream, scope.child_begin_position)? {
-            paint_scope(info, options, stream, &child_scope?, depth + 1, min_y)?;
+            let _ = paint_scope(info, options, stream, &child_scope?, depth + 1, min_y)?;
```
im not ignoring the result but the option. We return Ok(None) when the id does not have scope details, returning option because the error is about serialisation stuff and wasn't sure if its appropriated to add a new error there for when missing scope details for scope id.

TimonPost replied:
I always use clippy lint that enforced declaring a return type even if it isn't used. After this PR I'll add more lints to catch those cases.

TimonPost commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,248 @@
+    }
+}
+
+// Scopes are identified by user-provided name while functions are identified by the function name.
+#[derive(Debug, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+#[cfg_attr(
+    feature = "serialization",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub enum ScopeType {
+    /// The scope is a function profile scope identified by the name of this function.
+    Function(Cow<'static, str>),
+    /// The scope is a profile scope inside a function identified by the name of this scope.
+    Scope(Cow<'static, str>),
+}
```
I removed it, I did it this way mainly to make the hashmap lookup more clear without it scope name or function name are the hash keys. Two conceptually different things which could cause some confusion. I changed it to your suggestion think it works wel
2.

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+}
+
+// Scopes are identified by user-provided name while functions are identified by the function name.
```
```suggestion
/// Scopes are identified by user-provided name while functions are identified by the function name.
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+}
+
+impl ScopeType {
+    pub fn function_scope() -> Self {
+        Self::Function
+    }
+
+    pub fn scope() -> Self {
+        Self::Scope
+    }
+}
+
```
No need for these

```suggestion
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+    /// Returns what type of scope this is.
+    pub fn scope_type(&self) -> ScopeType {
+        // scope name is only set for non function scopes.
+        self.scope_name
+            .as_ref()
+            .map_or(ScopeType::function_scope(), |_| ScopeType::scope())
```
```suggestion
        if self.scope_name.is_some() { ScopeType::Scope } else { ScopeType::Function }
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+    /// The scope is a function profile scope generated by `puffin::profile_function!`.
+    Function,
+    /// The scope is a profile scope inside a function generated by `puffin::profile_scope!` or registered manually.
+    Scope,
```
`ScopeType::Scope` feels a bit redundant (aren't all scopes scopes?). I feel `ScopeType::Named` would be a better name

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    /// Returns the scope name if this is a profile scope or else the function name.
+    pub fn name(&self) -> &Cow<'static, str> {
+        self.scope_name.as_ref().map_or(&self.function_name, |x| x)
```
```suggestion
    #[inline]
    pub fn name(&self) -> &Cow<'static, str> {
        self.scope_name.as_ref().map_or(&self.function_name, |x| x)
```

possibly this could return a `&str` instead for simplicity

emilk commented on puffin_egui/src/flamegraph.rs:
```diff
@@ -731,19 +757,17 @@ fn paint_scope(
                     to_ms(scope.record.duration_ns)
                 ));
-                ui.monospace(format!("children: {num_children}"));
+                ui.monospace(format!("children: {num_children:3.3}"));
```
three decimals seems overkill for an integer ;)

```suggestion
                ui.monospace(format!("children: {num_children:3}"));
```

emilk commented on puffin/src/scope_details.rs:
```diff
@@ -0,0 +1,239 @@
+impl ScopeType {
+    pub fn type_str(&self) -> &'static str {
+        match self {
+            ScopeType::Function => "function scope",
+            ScopeType::Scope => "scope",
```
The ui currently reads "scope type: scope" which is just confusing

```suggestion
            ScopeType::Function => "function",
            ScopeType::Scope => "named",
```

emilk reviewed (approved):
Looks good!

repi reviewed (approved):
sweet

TimonPost commented:
Thanks @emilk for all the review comments! I'll merge this now, but won't do a release for about a week. Will maybe do some follow-ups and also update my own branch for Embarks internal project. Also, @MarijnS95 was working on migration and progressed some feedback from him in this PR. We can take it easy and see if any additional feature requests come in from concrete implementations. 

I also updated the migration help in the description, the changelog and the protocol version number.

emilk commented:
This broke saving to `.puffin` files :/
